### PR TITLE
proxy: record request Body for retry (fixes #1229)

### LINKED ISSUE #1229: proxy: retrying POST request failed with http: invalid Read on closed Body
(Are you asking for help with using Caddy? Please use our forum instead: https://forum.caddyserver.com. If you are filing a bug report, please answer the following questions. If your issue is not a bug report, you do not need to use this template. Either way, please consider donating if we've helped you. Thanks!)

### 1. What version of Caddy are you running (`caddy -version`)?
I tested the newest master caddy today.

### 2. What are you trying to do?
I try to test the proxy with POST method and body. When proxying to backend servers with load balance policy, if an error happened such as the backend server doesn't work well caddy would request another backend server. 
And I add one log in following location.
res, err := transport.RoundTrip(outreq)
	if err != nil {
		log.Println("err:", err)
		return err
	}

While if this failing request is a POST request and whose body has data, the second try would fail either because of the error "http: invalid Read on closed Body" even if the backend server is normal and it should be forwarded successfully in this second try.

### 3. What is your entire Caddyfile?
```text
localhost:2018

proxy / 192.168.206.231:9606 localhost:9608{
        policy round_robin
        max_fails 5
        try_interval 60s
}
```
PS: 192.168.206.231:9606 workes well while localhost:9608 doesn't exist.

### 4. How did you run Caddy (give the full command and describe the execution environment)?
OS: win7 x64
go version: go1.7
First I "go install" the source and then cd /bin directory, finally I run the caddy command
image

### 5. What did you expect to see?
I expect to see caddy can proxy successfully if I configure several endpoints even if there are some wrong endpoints.

### 6. What did you see instead (give full error messages and/or log)?
I got the log "err: http: invalid Read on closed Body" 
image

### 7. How can someone who is starting from scratch reproduce this behavior as minimally as possible?
I have Googled this issue and found an extremely similar one.  They have solved this.
https://github.com/flynn/flynn/issues/872

### PR DESCRIPTION
**Pull request by tpng**

### CODE REVIEW

**Comment by mholt** [comment]
Hey @tpng, thanks for the PR! Really glad you're helping out. :smile:

Did you see the fix for a similar issue in flynn? https://github.com/flynn/flynn/pull/875/files - they make "Close()" fake so that they only close it when they really mean to. What do you think of that approach instead of cloning? Seems to be a little simpler. (I haven't looked super thoroughly into the issue, but it seems like their problem was pretty much the same one.)

**Comment by tpng** [comment]
The reason for clone is that the request body maybe read by upstream already. In this case, fake close will end up with a dirty body for us to retry the request. It seems it is not possible for us to replay the downstream request without recording the body as it is read.

**Comment by tpng** [comment]
I think we don't need to clone it every retry, we can just record the downstream body once and reuse it.

**Comment by tpng** [comment]
@mholt PR updated.

File: caddyhttp/proxy/body.go [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+	return err
+}
+
+// newBody returns *body to use in place of src. Closes src and returns Read
+// error on src. All content from src is buffered in *body.
+func newBody(src io.ReadCloser) (*body, error) {
+	if src == nil {
+		return nil, nil
+	}
+	b, err := ioutil.ReadAll(src)
```
Comment (tw4452852) [inline caddyhttp/proxy/body.go]: I think we should close the original body here: `src.Close()`

File: caddyhttp/proxy/body.go [hunk 2]
```diff
@@ -0,0 +1,39 @@
+package proxy
+
+import (
+	"bytes"
+	"io"
+	"io/ioutil"
+)
+
+type body struct {
```
Comment (tw4452852) [inline caddyhttp/proxy/body.go]: Make a more sensitive name?

**Comment by tpng** [comment]
@tw4452852 Thanks! PR updated.

**Comment by tw4452852** [comment]
I think we'd better add a test case in `proxy_test.go` for issue #1229.

**Comment by tpng** [comment]
Any idea on how to write the test case for #1229?

**Comment by tpng** [comment]
@tw4452852 Test added for #1229.

**Comment by tw4452852** [comment]
Sorry for late. LGTM.

**Pull request merged**